src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs

```diff
@@ -358,6 +359,43 @@ internal static unsafe InnerSafeCloseSocket Accept(SafeSocketHandle socketHandle
                 res.SetHandle(acceptedFd);
                 return res;
             }
+
+            internal unsafe void TryUnblockSocket(bool abortive)
+            {
+                // Calling 'close' on a socket that has pending blocking calls (e.g. recv, send, accept, ...)
+                // may block indefinitly. This is a best effort attempt to not get blocked and make those operations return.
+                // We need to ensure we keep the expected TCP behavior that is observed by the socket peer (FIN vs RST close).
+                // What we do here isn't specified by POSIX, it works for Linux.
+
+                // Don't touch sockets which don't have the CLOEXEC flag set. These may be shared
+                // with other processes and we want to avoid disconnecting them.
+                int fdFlags = Interop.Sys.Fcntl.GetFD(this);
+                if (fdFlags == 0)
+                {
+                    return;
+                }
+
+                var linger = new Interop.Sys.LingerOption();
+                if (abortive ||
+                    (Interop.Sys.GetLingerOption(this, &linger) == Interop.Error.SUCCESS && linger.OnOff != 0))
+                {
+                    if (abortive || linger.Seconds == 0)
+                    {
+                        // Connect to AF_UNSPEC causes an abortive close (RST).
+                        Span<byte> address = stackalloc byte[32];
```

Comment:
This is going to contain garbage.  Does it need to be Clear'd?

Reply:
Only the address family will be checked.

Reply:
> Only the address family will be checked.

Why is that?  What prevents Connect from looking at the rest of it?

Reply:
Code always checks the address family and then reads the rest depending on the family. For AF_UNSPEC there is no rest.

Reply:
Do we really need to do the connect here? When I tested this while back, shutdown() seems to work for streams as well.